### PR TITLE
R interface for (de)serialization of rand_forest classifiers & regressors

### PR DESCRIPTION
**Pull request by yitao-li**
Signed-off-by: Yitao Li <yitao@rstudio.com>

### CODE REVIEW

File: tests/testthat/test-rand-forest-serde.R [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  actual_cls_probs <- predict_in_sub_proc(
+    model_state,
+    data = data,
+    expected_mode = "classification",
+    expected_model_cls = "cuda_ml_rand_forest",
+    additional_predict_args = list(output_class_probabilities = TRUE)
+  )
+
+  expect_equal(expected_preds, actual_preds)
+  expect_equal(expected_cls_probs, actual_cls_probs, tol = 1e-4, scale = 1)
```
Comment (yitao-li) [inline tests/testthat/test-rand-forest-serde.R:26]: At the moment RAPIDS cuML may have (I think) up to a single-precision ULP amount of rounding error introduced when reconstructing the rand forest model from the serialized state, which is why the new predictions are not exactly identical to the original ones, but are reasonably close.

**Comment by t-kalinowski** [comment]
Just tried to run tests, have you encountered this?
```R
==> devtools::test()

ℹ Loading cuda.ml

Attaching package: ‘magrittr’

The following objects are masked from ‘package:testthat’:

    equals, is_less_than, not


Attaching package: ‘rlang’

The following object is masked from ‘package:magrittr’:

    set_names

The following objects are masked from ‘package:testthat’:

    is_false, is_null, is_true

ℹ Testing cuda.ml
Error: The combination of 'cuda.ml' and mode 'classification' already has a fit component for model 'rand_forest'.
Backtrace:
     █
  1. └─devtools::test()
  2.   └─testthat::test_local(...)
  3.     └─testthat::test_dir(...)
  4.       └─testthat:::test_files(...)
  5.         └─testthat:::test_files(...)
  6.           └─testthat:::test_files_setup_env(...)
  7.             └─pkgload::load_all(test_dir, helpers = FALSE, quiet = TRUE)
  8.               └─pkgload:::run_pkg_hook(package, "load")
  9.                 └─nsenv[[f_name]](dirname(ns_path), package)
 10.                   └─cuda.ml:::register_rand_forest_model(pkgname) R/package.R:35:2
 11.                     └─parsnip::set_fit(...) R/rand_forest.R:487:4
Execution halted

Exited with status 1.
```

**Review by t-kalinowski — CHANGES_REQUESTED** [review]
Hi Yitao, the tests fail for me :(

**Comment by yitao-li** [comment]
@t-kalinowski Same here if launching the tests through `devtools`. I don't think we can make everything work with `devtools` when `{cuda.ml}` also does parsnip bindings.

I suggest instead of trying `devtools`, assign some key bindings to simply run the following from within RStudio

```
# There are probably more env variables you should set here if you want to emulate the env variable behavior of devtools, but really for testing purposes NOT_CRAN='true' is sufficient

cd tests

NOT_CRAN='true' Rscript testthat.R
```

**Comment by yitao-li** [comment]
@t-kalinowski I already tried. Deleting that line won't help though.

**Comment by yitao-li** [comment]
@t-kalinowski I think this problem probably would have been fixable if there is some way to unregister the `parsnip` bindings in `.onUnload()`, but I don't think `parsnip` supports unregistering at the moment.

**Comment by t-kalinowski** [comment]
FYI: 
```R
══ Warnings ════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
1. cuda_ml_kmeans() works as expected with 'kmeans++' initialization method (test-kmeans.R:25:3) - partial argument match of 'tol' to 'tolerance'

2. cuda_ml_kmeans() works as expected with 'random' initialization method (test-kmeans.R:36:3) - partial argument match of 'tol' to 'tolerance'

3. cuda_ml_kmeans() works as expected with user-specified initial cluster centers (test-kmeans.R:47:3) - partial argument match of 'tol' to 'tolerance'

4. random forest classifier can be serialized and unserialized correctly (test-rand-forest-serde.R:26:3) - partial argument match of 'tol' to 'tolerance'

5. random forest classifier can be serialized and unserialized correctly (test-rand-forest-serde.R:26:3) - partial argument match of 'tol' to 'tolerance'

6. random forest classifier can be serialized and unserialized correctly (test-rand-forest-serde.R:26:3) - partial argument match of 'tol' to 'tolerance'

7. random forest classifier can be serialized and unserialized correctly (test-rand-forest-serde.R:26:3) - partial argument match of 'tol' to 'tolerance'

8. random forest regressor can be serialized and unserialized correctly (test-rand-forest-serde.R:43:3) - partial argument match of 'tol' to 'tolerance'

9. random forest regressor can be serialized and unserialized correctly (test-rand-forest-serde.R:43:3) - partial argument match of 'tol' to 'tolerance'

══ DONE ════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
```
I have this snippet in my .Rprofile:
```r
} else { # not interactive

  options(
    warnPartialMatchArgs = TRUE,
    warnPartialMatchAttr = TRUE,
    warnPartialMatchDollar = TRUE
  )
```

**Review by t-kalinowski — APPROVED** [review]

**Pull request merged**